### ISSUE TITLE
Very intermittent core dump on FreeBSD x64

### ISSUE THREAD

**Issue by adam12** [boazsegev/iodine#40]
Not sure if you can debug from just the backtrace, but here it is. I normally only see it randomly after sending SIGINT, where it feels the process is completely not responding.

```
#0  0x0000000801e3584a in thr_kill () from /lib/libc.so.7
#1  0x0000000801e35814 in raise () from /lib/libc.so.7
#2  0x0000000801e35789 in abort () from /lib/libc.so.7
#3  0x0000000800a9005e in rb_bug () from /usr/local/lib/libruby24.so.24
#4  0x0000000800b5f1eb in ruby_sig_finalize () from /usr/local/lib/libruby24.so.24
#5  0x0000000801708934 in pthread_sigmask () from /lib/libthr.so.3
#6  0x0000000801707ecf in pthread_getspecific () from /lib/libthr.so.3
#7  <signal handler called>
#8  0x00000008056ab88b in sig_int_handler () from /usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so
#9  0x0000000801708926 in pthread_sigmask () from /lib/libthr.so.3
#10 0x0000000801707ecf in pthread_getspecific () from /lib/libthr.so.3
#11 <signal handler called>
#12 0x0000000801eb6a8a in munmap () from /lib/libc.so.7
#13 0x00000008056a760a in facil_libcleanup () from /usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so
#14 0x0000000801ea44d5 in __cxa_finalize () from /lib/libc.so.7
#15 0x0000000801e353d1 in exit () from /lib/libc.so.7
#16 0x00000008056ab995 in facil_sentinel_worker_thread () from /usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so
#17 0x00000008056eb1ab in defer_thread_start () from /usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so
#18 0x0000000800b9403e in rb_thread_call_without_gvl () from /usr/local/lib/libruby24.so.24
#19 0x00000008056eb17a in defer_thread_inGVL () from /usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so
#20 0x0000000800b9a0ef in rb_reset_coverages () from /usr/local/lib/libruby24.so.24
#21 0x0000000800b99da5 in rb_reset_coverages () from /usr/local/lib/libruby24.so.24
#22 0x0000000801702bc5 in pthread_create () from /lib/libthr.so.3
#23 0x0000000000000000 in ?? ()
```

`FreeBSD  11.2-RELEASE FreeBSD 11.2-RELEASE #0 r335510: Fri Jun 22 04:32:14 UTC 2018     root@releng2.nyi.freebsd.org:/usr/obj/usr/src/sys/GENERIC  amd64`

`ruby 2.4.4p296 (2018-03-28 revision 63013) [amd64-freebsd11]`

Let me know if there's something better to debug against (core perhaps?) or if I can do anything to help debug this.

Much thanks!

**Comment by boazsegev** [comment]
Hi @adam12

Thank you for opening this issue 🙏🏻

Technically, it's obviously a bug. However, it appears to be very low priority since **it doesn't effect the live application**.

It's hard to follow from the backtrace, but it's clear the issue occurs somewhere during `facil_libcleanup`... 

...which means that the issue occurs after the server completed a clean shutdown and conceivably after the Ruby layer was freed (depending on the system's `atexit` ordering scheme).

Which iodine version are you using? Does this happen when using version 0.6.5?

Again, thank you so much for opening the issue. I hope I can resolve the issue soon.

Kindly,
Bo.

**Comment by boazsegev** [comment]
P.S.

I'm more worried about this:

> I normally only see it randomly after sending SIGINT, **where it feels the process is completely not responding**.

Is it possible to re-create this "hanging" and find it's roots? Anything that effects the live application is a high priority concern for iodine.

B.

**Comment by adam12** [comment]
> Technically, it's obviously a bug. However, it appears to be very low priority since it doesn't effect the live application.

Completely agree :+1:

> Which iodine version are you using? Does this happen when using version 0.6.5?

I've seen it occur on a few versions, 0.6.5 included (and for sure 0.6.2), since a few months ago when I started using Iodine.

> ...which means that the issue occurs after the server completed a clean shutdown and conceivably after the Ruby layer was freed (depending on the system's atexit ordering scheme).

> Is it possible to re-create this "hanging" and find it's roots? Anything that effects the live application is a high priority concern for iodine.

To be clear, the application is working correctly when I send `SIGINT`. It's only during the cleanup phase (when `* $PID cleaning up` is logged) that it feels like it becomes unresponsive. At that point, I'm assuming `SIGINT` is no longer trapped correctly and the only way to abort the process is `SIGKILL`. It seems to coredump around this time.

The reality is that this is just more a nuisance than a concern, but I figured I'd report it anyways. Cheers :)

**Comment by adam12** [comment]
Forgot I had this available still. You can see where I send `^C`, and Iodine performing the shutdown routine and then the segfault.

```
* Listening on port 3000

Starting up Iodine:
 * Ruby v.2.4.4
 * Iodine v.0.6.5
 * 2 Workers X 2 Threads per worker.

Server is running 2 workers X 2 threads, press ^C to stop
* Detected capacity: 57987 open file limit
* Root pid: 48657
* 48659 is running.
* 48660 is running.

^C^[[A* (48660) Detected exit signal.
* Server Detected exit signal.
* (48659) Detected exit signal.
* Stopped listening on port 3000
* 48657 cleaning up.

* 48659 cleaning up.
* 48660 cleaning up.
[BUG] Segmentation fault at 0x0000000800677004
ruby 2.4.4p296 (2018-03-28 revision 63013) [amd64-freebsd11]

-- Control frame information -----------------------------------------------
c:0001 p:---- s:0003 e:000002 (none) [FINISH]


-- C level backtrace information -------------------------------------------
/usr/local/lib/libruby24.so.24 [0x800be034f]
/usr/local/lib/libruby24.so.24 [0x800a90005]
/usr/local/lib/libruby24.so.24 [0x800b5f1eb]
/lib/libthr.so.3 [0x801708934]
/lib/libthr.so.3 [0x801707ecf]
[0x7ffffffff193]
/usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so(sig_int_handler+0x2b) [0x8056ab88b]
/lib/libthr.so.3 [0x801708926]
/lib/libthr.so.3 [0x801707ecf]
[0x7ffffffff193]
/lib/libc.so.7(_munmap+0xa) [0x801eb6a8a]
/usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so(facil_each+0xea) [0x8056a760a]
/lib/libc.so.7(__cxa_finalize+0x145) [0x801ea44d5]
/lib/libc.so.7(exit+0x21) [0x801e353d1]
/usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so(facil_each+0xf5) [0x8056ab995]
/usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so(iodine_defer_initialize+0x1b) [0x8056eb1ab]
/usr/local/lib/libruby24.so.24 [0x800b9403e]
/usr/local/lib/ruby/gems/2.4/gems/iodine-0.6.5/lib/iodine/iodine.so(iodine_defer_initialize+0x4a) [0x8056eb17a]
/usr/local/lib/libruby24.so.24 [0x800b9a0ef]
/usr/local/lib/libruby24.so.24 [0x800b99da5]
/lib/libthr.so.3 [0x801702bc5]
[0x0]
```

**Comment by adam12** [comment]
When it enters the failed shutdown state, and appears hung, I was able to attach `truss` to see what was kinda going on.

Root PID (50014) was doing zero syscalls but still running. I'm assuming it was waiting on `waitpid` to return but it's not logged.
One of the workers (pid 50016) was in an infinite `nanosleep` loop. `nanosleep({ 1.000000000 })                       = 0 (0x0)`
The other worker (pid 50017) was shutdown.

In the interest of being detailed (but I don't think this matters as it's likely just due to the scheduling of the threads), I also noticed that the log order is possibly off. We see 50016 log running, then logging from the root pid, and finally a log from 50017 (the one that properly shut down). Is there a chance that these processes have started out of order and 50016 ended up being orphaned somehow? I say this hypothetically because according to the pstree, the root pid definitely owns 50016. None of them are zombies.

```
* Listening on port 3000

Starting up Iodine:
 * Ruby v.2.4.4
 * Iodine v.0.6.5
 * 2 Workers X 2 Threads per worker.

* 50016 is running.
Server is running 2 workers X 2 threads, press ^C to stop
* Detected capacity: 57987 open file limit
* Root pid: 50014
* 50017 is running.
```

To add to this, I've seen `pg` complain about messages coming in after being expected. I think this is because a pg connection was taken across `fork`. I've implemented a call to `Iodine.before_fork` which might be the entire root cause of this? so I'll report back after I've spent a bit of time with it.

**In summary: maybe this is my own fault. Will advise. :grin:**

**Comment by boazsegev** [comment]
Wow, great research, thanks!

I’m writing on my phone, so I’ll be short. I’ll explore and explain more when I’m next to my computer.

The infinite loop is probably a spin-lock. I would love to know which function is spinning, though it might be hard to track down unless it’s a critical lock (or a third party lock) that doesn’t yield control to the event loop. 

The root process will be waiting for all the children to shutdown. Since this is it’s last task, it’s clearly not affected by the issue.

The children/workers run concurrently, so it’s not surprising that the messages are out of order. Also, I think the root process will forward the exit signal before reporting, partly because I can’t control the locks on stderr.

As for forking, it’s important to close any database connections before forking and re-establish them afterwards (which is often lazily performed, so the before_fork should work... but maybe it doesn’t, maybe the error occurs on startup and is only observable during shutdown. Is the CPU spinning while the app is live?

b.t.w...

Do you have a good vagrant file that would allow me to test locally with the same environment?

B.

**Comment by adam12** [comment]
> As for forking, it’s important to close any database connections before forking and re-establish them afterwards (which is often lazily performed, so the before_fork should work... but maybe it doesn’t, maybe the error occurs on startup and is only observable during shutdown

I'm starting to believe this is more and more the culprit. Although not 100% sure, since I added a DB disconnect in the `before_fork` hook, I haven't seen the deadlock. I thought it might have been Iodine related due to the backtrace but could of just manifested in Iodine's cleanup process.

> Is the CPU spinning while the app is live?

Not that I can see.

> Do you have a good vagrant file that would allow me to test locally with the same environment?

Afraid not. I'm just using the FreeBSD cloud image, running in plain VirtualBox. [This image ](https://download.freebsd.org/ftp/releases/VM-IMAGES/11.2-RELEASE/amd64/Latest/FreeBSD-11.2-RELEASE-amd64.vhd.xz) or one of the other ones from [here](https://download.freebsd.org/ftp/releases/VM-IMAGES/11.2-RELEASE/amd64/Latest/).

**Comment by boazsegev** [comment]
> ...since I added a DB disconnect in the before_fork hook, I haven't seen the deadlock...

Interesting.

Does this mean that disconnecting from `pg` solved the issue?

Wish I could track down the conflict and discover why that is.

### As a sidenote:

I noticed this line in the log: `2 Workers X 2 Threads per worker.`

This make me assume that the machine has 2 CPU cores and that no concurrency optimizations were attempted...?

Depending on the application's design, I would consider increasing the thread count.

With the `pg` gem clearly part of the stack, threads might be blocked on IO by the `pg` gem. This would mean that increasing the thread count (for example, using 16 threads per worker), could improve performance.

The more blocking IO calls in your application, the more threads you could add to make up for the "waiting" cycles.

Try updating the command line to:

     $ bundle exec iodine -w -1 -t 16 -www ./public 

This will translates to :

* `-1` workers, which is equal to Cores / 1  (`-2` == Cores / 2).
* 16 threads per worker
* static file service from the folder `./public` (which also adds `X-Sendfile` header support and automatic `.gz` pre-compressed resource detection).

You could see more options (such as maximum upload size, ping intervals etc') running `iodine -?`. Note that Unix Sockets are supported (and often preferred when using nginx or apache).

**Comment by adam12** [comment]
I'm on a different project without Iodine now, but I didn't see it re-occur after adding the `before_fork` hook (tho that might be co-incidence).

I should be back on Iodine in a few weeks, so I'm going to close this and I'll re-open it if I can reproduce the `before_fork` hook solving the issue. Let's presume it's at the application level and not the server level.

Cheers.
